source/ScrollBar.cpp

```diff
@@ -0,0 +1,170 @@
+/* ScrollBar.cpp
+Copyright (c) 2024 by Daniel Yoon
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "ScrollBar.h"
+
+#include "LineShader.h"
+#include "PointerShader.h"
+
+#include <algorithm>
+
+namespace {
+	// Additional distance the scrollbar's tab can be selected from.
+	constexpr int SCROLLBAR_MOUSE_ADDITIONAL_RANGE = 5;
+}
+
+
+
+ScrollBar::ScrollBar(
+	float fraction,
+	float displaySizeFraction,
+	const Point &from,
+	const Point &to,
+	float tabWidth,
+	float lineWidth,
+	Color color,
+	Color innerColor
+) noexcept :
+	fraction(fraction),
+	displaySizeFraction(displaySizeFraction),
+	from(from),
+	to(to),
+	tabWidth(tabWidth),
+	lineWidth(lineWidth),
+	color(color),
+	innerColor(innerColor)
+{
+}
+
+
+
+ScrollBar::ScrollBar() noexcept : ScrollBar(
+	0.,
+	0.,
+	Point(),
+	Point(),
+	3,
+	3,
+	Color(0.6),
+	Color(0.25)
+)
+{
+}
+
+
+
+void ScrollBar::Draw()
+{
+	DrawAt(from);
+}
+
+
+
+void ScrollBar::DrawAt(const Point &from)
+{
+	Point delta = to - this->from;
+	LineShader::Draw(
+		from,
+		from + delta,
+		lineWidth,
+		innerHighlighted ? innerColor : Color::Multiply(.5, innerColor)
+	);
+
+	Point deltaOffset = delta * displaySizeFraction;
+	Point offset = delta * (1 - displaySizeFraction) * fraction;
+	LineShader::Draw(
+		from + offset + deltaOffset,
+		from + offset,
+		tabWidth,
+		highlighted ? color : Color::Combine(.5, color, .5, innerColor)
+	);
+
+	PointerShader::Draw(from, Point(0., -1.), lineWidth * 3., 10.f, 5.f, fraction > 0. ? color : innerColor);
+	PointerShader::Draw(from + delta, Point(0., 1.), lineWidth * 3., 10.f, 5.f, fraction < 1. ? color : innerColor);
+}
+
+
+
+bool ScrollBar::Hover(int x, int y)
+{
+	Point delta = to - from;
+	Point offset = delta * (1.0 - displaySizeFraction) * fraction;
+
+	constexpr auto LineSDF = [](Point a, Point b, Point p)
```

Comment:
Signed distance function? This could use a comment. ~~Or just rename it to LineDistance~~